File: ShearNDipityLib/src/dao/DAO.java
package dao;

import dao.restDAO.RESTFactory;

public class DAO implements IDAOFactory {
    private static DAO instance;
    private IDAOFactory factory;

    protected DAO() {
        factory = new RESTFactory();
    }

    public static DAO getInstance() {
        if (instance == null) {
            instance = new DAO();
        }
        return instance;
    }

    private IContactDAO getContactsFactoryDAO() {
        return factory.getContactDAO();
    }

    private IUserDAO getUserFactoryDAO() {
        return factory.getUserDAO();
    }

    private IProviderDAO getProviderFactoryDAO() {
        return factory.getProviderDAO();
    }

    private ITypeDAO getTypeFactoryDAO() {
        return factory.getTypeDAO();
    }

    private IAvailabilitiesDAO getAvailabilitiesFactoryDAO() {
        return factory.getAvailabilitiesDAO();
    }

    private IAppointmentDAO getAppointmentFactoryDAO() {
        return factory.getAppointmentDAO();
    }

    public ITypeDAO getTypeDAO() {
        return getInstance().getTypeFactoryDAO();
    }

    public IContactDAO getContactDAO() {
        return getInstance().getContactsFactoryDAO();
    }

    public IUserDAO getUserDAO() {
        return getInstance().getUserFactoryDAO();
    }

    public IProviderDAO getProviderDAO() {
        return getInstance().getProviderFactoryDAO();
    }

    public IAvailabilitiesDAO getAvailabilitiesDAO() {
        return getInstance().getAvailabilitiesFactoryDAO();
    }

    public IAppointmentDAO getAppointmentDAO() {
        return getInstance().getAppointmentFactoryDAO();
    }
}

File: Server/server/dao/DAO.js
var instance;

var Ring = require('ring');
var IDAOFactory = require('./IDAOFactory.js');

var DAO = Ring.create({
    setFactory: function (factory) {
        if (!Ring.instance(factory, IDAOFactory)) {
            throw new Error('Invalid Factory Passed.');
        } else {
            this._factory = factory;
        }
    },
    getFactory: function () {
        return this._factory;
    }
});

module.exports = {
    getInstance: function () {
        return instance || (instance = new DAO());
    },
    getAppointmentTypeDAO: function () {
        return this.getInstance().getFactory().getAppointmentTypeDAO();
    },
    getContactDAO: function () {
        return this.getInstance().getFactory().getUserDAO();
    },
    getUserDAO: function () {
        return this.getInstance().getFactory().getUserDAO();
    },
    getProviderDAO: function () {
        return this.getInstance().getFactory().getProviderDAO();
    }
};



